### PR TITLE
RFC: Types

### PR DESCRIPTION
**Pull request by CaptainHayashi**
It might be useful to describe Bifrost's type system.  I've started an initial RFC here (**DO NOT MERGE**), so we can discuss over the lines:

* More types or fewer types?
* 64-bit int type?  Should our int type be 64-bit anyway?
* Float type?  URL type?  The possibilities are endless.
* Should directories themselves be typed (playlists being `directory<directory<string>>` or just `directory`)?
* Distinct integer-indexed-list type?
* If directories have special semantics for writing and deleting, should they still be directories or should they be a new type?
* Are mounts worth it?
* Should the query value have a different format?
* etc.

### CODE REVIEW

File: protocol/types.md [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+These types are more specialised.
+
+### Query
+
+A query represents some external process, such as a database query, where the set of values is not known up-front.
+Queries are read-only, and any attempt to read causes that external process to run with the path into the query
+given as input.
+
+The value of a query is a schema, of the form `{type};{field1}/{field2}/.../{fieldN}`,
```
Comment (LordAro) [inline protocol/types.md]: This seems a little odd to me, but i guess if it works.. is it used in trackd?

File: protocol/types.md [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+that describes the return type of the query and the format of valid URLs.
+For example, a search for tracks by artist and title may look like `directory;{artist}/{title}`.
```
Comment (LordAro) [inline protocol/types.md]: This doesn't seem to be an actual example, more like an example of what an example would look like

**Comment by LordAro** [comment]
> More types or fewer types?
> Float type? URL type? The possibilities are endless.

Shouldn't be adding types that we're not using

> 64-bit int type? Should our int type be 64-bit anyway?

Possibly (how large are the TIME (equivalents) going to be?). I think it's easier if we only have 1 integer type used though (simpler, no signedness issues)

> Should directories themselves be typed (playlists being directory<directory<string>> or just directory)?

Seems like nested types would just make everything more complicated. Might conflict with the mounts idea as well?

> Distinct integer-indexed-list type?

Nah

>If directories have special semantics for writing and deleting, should they still be directories or should they be a new type?

>Are mounts worth it?

Possibly.

>Should the query value have a different format?

See comment. If it works, I think it'll be fine. For instance, how are different keys specified? i.e. only searching for title

File: protocol/types.md (hunk 1 above)
Comment (CaptainHayashi) [inline protocol/types.md]: Not yet, but I'm not too keen on the idea of a `directory` that says it has no contents and behaves somewhat weirdly when you read it.

**Comment by CaptainHayashi** [comment]
> Possibly (how large are the TIME (equivalents) going to be?). I think it's easier if we only have 1 integer type used though (simpler, no signedness issues)

Signed 64-bit integer might be good enough.  If anything, we can use strings for anything else (lol, JSON).

> Seems like nested types would just make everything more complicated. Might conflict with the mounts idea as well?

Yeah, let's not do it.  I do wonder if it's worth writing in the RFC that clients should slice anything after a / out of the type name just to provide extensibility in case we do do more complicated shit... but maybe this would be a major protocol revision thing instead.

> See comment. If it works, I think it'll be fine. For instance, how are different keys specified? i.e. only searching for title

Not sure, unless we define a more specialised language for query descriptions... or they're separate query endpoints.

**Pull request merged**